src/pybind/mgr/dashboard/services/ganesha.py

```diff
@@ -0,0 +1,894 @@
+
+        for t in self.transports:
+            match = re.match(r'^TCP$|^UDP$', t)
+            if not match:
+                raise NFSException(
+                    "'{}' is an invalid network transport type identifier"
+                    .format(t))
+
+        self.fsal.validate()
+
+        if 4 in self.protocols:
+            if not self.pseudo:
+                raise NFSException(
+                    "Pseudo path is required when NFSv4 protocol is used")
+            match = re.match(r'^/[^><|&()]*$', self.pseudo)
+            if not match:
+                raise NFSException(
+                    "Export pseudo path ({}) is invalid".format(self.pseudo))
+
+        if self.tag:
+            match = re.match(r'^[^/><|:&()]+$', self.tag)
+            if not match:
+                raise NFSException(
+                    "Export tag ({}) is invalid".format(self.tag))
+
+        if self.fsal.name == 'RGW' and 4 not in self.protocols and not self.tag:
+            raise NFSException(
+                "Tag is mandatory for RGW export when using only NFSv3")
+
+    @classmethod
+    def from_export_block(cls, export_block, defaults):
+        logger.debug("[NFS] parsing export block: %s", export_block)
+
+        fsal_block = [b for b in export_block['_blocks_']
+                      if b['block_name'] == "FSAL"]
+
+        protocols = export_block.get('protocols', defaults['protocols'])
+        if not isinstance(protocols, list):
+            protocols = [protocols]
+
+        transports = export_block.get('transports', defaults['transports'])
+        if not isinstance(transports, list):
+            transports = [transports]
+
+        client_blocks = [b for b in export_block['_blocks_']
+                         if b['block_name'] == "CLIENT"]
+
+        return cls(export_block['export_id'],
+                   export_block['path'],
+                   FSal.from_fsal_block(fsal_block[0]),
+                   [],
+                   export_block.get('pseudo', None),
+                   export_block.get('tag', None),
+                   export_block.get('access_type', defaults['access_type']),
+                   export_block.get('squash', defaults['squash']),
+                   protocols,
+                   transports,
+                   [Client.from_client_block(client)
+                    for client in client_blocks])
+
+    def to_export_block(self, defaults):
+        result = {
+            'block_name': 'EXPORT',
+            'export_id': self.export_id,
+            'path': self.path
+        }
+        if self.pseudo:
+            result['pseudo'] = self.pseudo
+        if self.tag:
+            result['tag'] = self.tag
+        if 'access_type' not in defaults \
+                or self.access_type != defaults['access_type']:
+            result['access_type'] = self.access_type
+        if 'squash' not in defaults or self.squash != defaults['squash']:
+            result['squash'] = self.squash
```

Comment:
One more thing I think we want here. Ganesha can support labeled NFS on FSAL_CEPH. In an export block, you can set:

`security_label = true;`

Maybe a new field for the json here, and allow this to be enabled (with the default being false)?

The default is to use the xattr "security.selinux", but there is a setting in FSAL_CEPH to switch that to a different one if that's desirable (i.e. if someone is using SMACK instead of SELinux). We could also expose that field too if there is interest in alternate LSMs.

Reply:
Sure, adding fields is easy. Better to add them now than later.

Also, to keep saved somewhere before I forget, we have discussed the inclusion of `Attr_Expiration_Time = 0;` in all cephfs exports.

Reply:
@jtlayton what are the possible values for the fsal option "sec_label_xattr"?

Reply:
It's a free-form string. I'd probably just allow the admin to override it with whatever they like, but maybe make note in the webui that the default is "security.selinux".

Reply:
@jtlayton one last question, if `security_label == False` then we can ignore the `sec_label_xattr`, correct?

Reply:
Correct.